include/libcyphal/application/registry/registry_impl.hpp

```diff
@@ -0,0 +1,215 @@
+#define LIBCYPHAL_APPLICATION_REGISTRY_IMPL_HPP_INCLUDED
+
+#include "libcyphal/common/cavl/cavl.hpp"
+#include "registry.hpp"
+#include "registry_value.hpp"
+
+#include <cetl/cetl.hpp>
+#include <cetl/pf17/cetlpf.hpp>
+
+namespace libcyphal
+{
+namespace application
+{
+namespace registry
+{
+
+/// Defines the registry implementation.
+///
+class Registry final : public IIntrospectableRegistry
+{
+public:
+    /// Constructs a new registry.
+    ///
+    /// @param memory The memory resource to use for variable size-d register values.
+    ///
+    explicit Registry(cetl::pmr::memory_resource& memory)
+        : memory_{memory}
+    {
+    }
+    ~Registry() = default;
+
+    Registry(Registry&&)                 = delete;
+    Registry(const Registry&)            = delete;
+    Registry& operator=(Registry&&)      = delete;
+    Registry& operator=(const Registry&) = delete;
+
+    cetl::pmr::memory_resource& memory() const noexcept
+    {
+        return memory_;
+    }
+
+    // MARK: - IRegistry
+
+    cetl::optional<IRegister::ValueAndFlags> get(const IRegister::Name name) const override
+    {
+        if (const auto* const reg = findRegisterBy(name))
+        {
+            return reg->get();
+        }
+        return cetl::nullopt;
+    }
+
+    cetl::optional<SetError> set(const IRegister::Name name, const Value& new_value) override
+    {
+        if (auto* const reg = findRegisterBy(name))
+        {
+            return reg->set(new_value);
+        }
+        return SetError::Existence;
+    }
+
+    // MARK: - IIntrospectableRegistry
+
+    std::size_t size() const override
+    {
+        return registers_tree_.size();
+    }
+
+    IRegister::Name index(const std::size_t index) const override
+    {
+        if (const auto* const reg = registers_tree_[index])
+        {
+            return reg->getName();
+        }
+        return IRegister::Name{};
+    }
+
+    bool append(IRegister& reg) override
+    {
+        CETL_DEBUG_ASSERT(!reg.isLinked(), "Should not be linked yet.");
+
+        auto register_existing = registers_tree_.search(
+            [key = reg.getKey()](const IRegister& other) {
+                //
+                return other.compareBy(key);
+            },
+            [&reg]() -> IRegister* {
+                //
+                return &reg;
+            });
+
+        CETL_DEBUG_ASSERT(std::get<0>(register_existing) != nullptr, "");
+        CETL_DEBUG_ASSERT(std::get<0>(register_existing)->isLinked(), "Should be linked.");
+        return !std::get<1>(register_existing);
+    }
+
+    // MARK: - Other factory methods:
+
+    /// Constructs a new read-only register, and links it to this registry.
+    ///
+    /// @param name The name of the register. Should be unique within the registry.
+    /// @param options Extra options for the register, like "persistent" option (`true` by default).
+    /// @param getter The getter function to provide the register value (from a storage).
```

Comment:
```suggestion
    /// @param getter The getter function to provide the register value.
```

There may be no storage at all. It is common to construct values ad hoc.